group enable functionality

SeojLee opened the pull request:
<!-- kody-pr-summary:start -->
This pull request introduces a new group pausing (freeze) functionality, deprecates GSI authentication for newer HTCondor versions, and removes the `cacertificate` field from cloud configurations.

Key changes include:

*   **Group Pausing/Unpausing (Freeze functionality):**
    *   Implemented automatic pausing (freezing) of groups if their associated HTCondor poller is detected as inactive for a configurable timeout.
    *   Groups are automatically unpaused (unfrozen) when their HTCondor poller becomes active again.
    *   VM provisioning is now skipped for groups that are in a paused state.
    *   A "Paused" checkbox has been added to the group defaults page in the web interface, allowing administrators to manually pause or unpause groups.
    *   New `freeze` columns have been added to the `csv2_groups` and `csv2_jobs` tables, and the group freeze status is now included in the `view_idle_vms` view.

*   **GSI Authentication Deprecation for HTCondor 9+:**
    *   The `condor_poller.py` now detects HTCondor versions 9.x and later, automatically deprecating GSI authentication for these versions.
    *   GSI-related credentials (`htcondor_gsi_dn`, `htcondor_gsi_eol`, `worker_dn`, `worker_eol`, `worker_cert`, `worker_key`) are automatically cleared from the database for hosts running deprecated GSI authentication.

*   **Removal of `cacertificate` Field:**
    *   The `cacertificate` field has been removed from the cloud configuration database schema, documentation, and all relevant web UI forms (add/edit cloud).

*   **Minor Enhancements and Fixes:**
    *   Added new database fields: `retire`, `terminate`, `LoadAvg`, `Memory` to `csv2_vms`, `visible` to `csv2_service_catalog`, and `show_status` to `view_vms` for enhanced monitoring and status reporting.
    *   The status page refresh button now toggles refresh on/off instead of just stopping it.
    *   Fixed URL encoding for InfluxDB queries on the status page.
    *   Adjusted default value handling for `user_domain_name` and `project_domain_name` in cloud configuration forms.
    *   Corrected column sorting indices in the VM list table in the web UI.
<!-- kody-pr-summary:end -->

SeojLee committed:
removed cacert field

SeojLee committed:
removed cacertificate from CLI tool

SeojLee committed:
fixed indexing

SeojLee committed:
removed gsi ERROR message for HTCondor 9 and beyond

SeojLee committed:
fixed issue 390

SeojLee committed:
-logs

SeojLee committed:
added toggle refresh functionality

SeojLee committed:
Remove cacertificate from view_clouds

SeojLee committed:
Remove cacertificate field from csv2_clouds.yaml

SeojLee committed:
Remove cacertificate parameter from view_clouds.yaml

SeojLee committed:
Remove cacertificate field from view_clouds schema

SeojLee committed:
Simplify session establishment by removing CA cert check

SeojLee committed:
full functionality complete

SeojLee committed:
Delete cloudscheduler/csmaincp

SeojLee committed:
moved check to main process

SeojLee committed:
Clean up blank lines in GSI poller functions

Removed unnecessary blank lines in worker and condor GSI poller functions.

SeojLee committed:
Refactor GSI database clearing and logging

Updated logging message to be more general and removed deprecated process IDs for HTCondor versions 9 and above.

SeojLee committed:
revised freeze/unfreeze functionality to mirror cloud freeze/unfreeze

colsond commented on schema/schema_backup/elephant190.heprc.uvic.ca:
we dont want this backup in the dev/production branches

colsond commented on schema/views/view_starting_vms:
This change is not used in this pull request and should be reverted.

SeojLee committed:
Refactor resource contention checks for VM booting

SeojLee committed:
Add user and project domain name fields to clouds.html

SeojLee committed:
Refactor HTML for user and project domain inputs

SeojLee committed:
Refactor view_starting_vms SQL query

colsond commented:
In addition to stopping VM scheduling when a condor poller is inactive, we also want to stop flagging retires/kills for the group in question. This likely means updating the view_idle_vms to include the freeze status for a group.

SeojLee committed:
logging tests

SeojLee committed:
Full functionality of skip retire idle when group frozen

SeojLee committed:
Delete select

colsond commented on cloudscheduler/csmain:
```diff
@@ -253,6 +253,9 @@ def check_view_idle_vms():
             rc, msg, results = config.db_query("view_idle_vms", where=where_clause)
             vmids = []
             for res in results:
+                if len(results)>0 and res['freeze']== 1:
```
move to where clause

colsond commented on cloudscheduler/csmain:
```diff
@@ -556,6 +559,83 @@ def process_group_cloud_terminates(pair, config):
 
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+def get_groups_with_inactive_pollers(config, timeout_seconds=10):
```
Make the timeout configurable with a maximum of 10 minutes to avoid thrashing if the poller is down.

colsond commented:
Add a indicator for group pause in the group column. Make the cell yellow when its paused and then expand it for more information (automatic vs manual pause).

colsond commented on cloudscheduler/csmain:
```diff
@@ -655,6 +744,13 @@ def main():
             if list_groups_with_idle_jobs_len > 0:
                 for job_group in list_groups_with_idle_jobs[list_groups_with_idle_jobs_ix:] + list_groups_with_idle_jobs[:list_groups_with_idle_jobs_ix]:
+                    
+                    where_clause = f"group_name='{job_group}'"
```
WE can avoid this query by updating the view to either include freeze or exclude any groups with freeze set.

colsond commented on data_collectors/condor/condor_poller.py:
need to update the retire function//query to include the pause functionality.

colsond replied:
We could update view_condor_host to include freeze so we can filter them out during the query step.

colsond replied:
Main also references view_condor_host at line 311 when doing terminates and will also need a where_clause update.

SeojLee committed:
Remove frozen VM check in retirement process

Removed check for frozen VMs before appending to vmids.

SeojLee committed:
Update view_groups_of_idle_jobs

SeojLee committed:
Update view_groups_of_idle_jobs

SeojLee committed:
Delete schema/schema_backup/elephant190.heprc.uvic.ca

SeojLee committed:
Skip retiring frozen VMs in csmain logic

Add check to skip retiring frozen VMs in csmain.

SeojLee committed:
Remove inactive condor poller check for job groups

Removed code that checks for inactive condor poller before provisioning job groups.

SeojLee committed:
Refactor get_groups_with_inactive_pollers and get_frozen_groups_with_active_pollers

Updated functions to remove timeout_seconds parameter and handle its value internally based on configuration.

SeojLee committed:
Remove log statement for active pollers

Remove logging of poller information in active pollers loop.

SeojLee committed:
Add freeze condition to where clause in query

SeojLee committed:
Fix where clause in get_groups_name function

SeojLee committed:
Correct indentation for timeout_seconds assignment

Fix indentation for setting timeout_seconds to 600.

SeojLee committed:
Add 'condor_poller.py' to config parameters

Commit:
added exclusion to view_condor_host and added group disable symbol and colour coding

Commit:
removed files

SeojLee committed:
Delete web_frontend/cloudscheduler/wsgi.py

SeojLee committed:
Delete schema/schema_backup/elephant178.heprc.uvic.ca

SeojLee committed:
Update cloud_views.py

SeojLee committed:
Refactor group_defaults.html for cleaner structure

SeojLee committed:
Refactor HTML for job freeze status display

SeojLee committed:
Change freeze condition to handle null values

colsond committed:
Merge branch 'stable-2.11' into seojin-group-enable
